DAOS 5066 daos: add cont clone option to utilities

dsikich opened the pull request:
Add generic container clone utility

DAOS clone utility that will clone all types of containers.
This utility supports passing in a pool and container as
well as using a Unified Namespace Path. Currently, the
utility assumes that the source pool and cont already exists,
and just the pool for the destination. This utility only supports 
DAOS -> DAOS copies, and does not support moving data
to and from POSIX. The container clone utility does not
support data movement to already existing destination
containers. If a new unified namespace path is given, then
the destination container will use the source pool. 

example: daos cont clone --src=/$pool1/$p1cont1 --dst=/$pool1

example with UNS: daos cont clone --src=/tmp/$USER/conts/uns0
		                                             --dst=/tmp/$USER/conts/newuns

Signed-off-by: Danielle Sikich <danielle.sikich@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/1/execution/node/334/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/1/execution/node/263/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/1/execution/node/296/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

dsikich commented:
@johannlombardi one of the builds caught an error where I missed returning the length on one of the snprintf's I was using, so I needed to update.

daltonbohning commented on src/utils/daos.c:
```diff
@@ -644,17 +644,17 @@ common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
-				D_GOTO(out_free, rc = RC_NO_HELP);
-			}
-			break;
+			D_GOTO(out_free, rc = RC_NO_HELP);
+		}
+		break;
```
@dsikich Formatting:
Looks like a tab was removed for these lines

daltonbohning commented on src/utils/daos.c:
```diff
@@ -644,17 +644,17 @@ common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
-				D_GOTO(out_free, rc = RC_NO_HELP);
-			}
-			break;
+			D_GOTO(out_free, rc = RC_NO_HELP);
+		}
+		break;
```
@dsikich Formatting:
Same, tabs were removed

daltonbohning commented on src/utils/daos.c:
```diff
@@ -1091,6 +1108,12 @@ cont_op_hdlr(struct cmd_args_s *ap)
+		break;
+
+
+	/* TODO: implement the following ops */
```
@dsikich Question:
Should this TODO have been added? Isn't this implemented?

dsikich replied:
@daltonbohning I don't know that was already there. I did not add that TODO.

daltonbohning replied:
@dsikich I **think** that TODO was there before you started this branch, but then it was resolved _**during**_ your development of this branch. Maybe a merge accidentally put it back in? I do believe `daos obj list` works now, and in _**master**_ this TODO is not there

daltonbohning replied:
For reference, in master:
https://github.com/daos-stack/daos/blob/130088667ac84bde8437df97027e13518306adf9/src/utils/daos.c#L1091-L1097

dsikich replied:
@daltonbohning ok, I removed it. It must have stayed there because I added the CONT_COPY.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2505,11 +2508,14 @@ fs_copy_connect(struct file_dfs *src_file_dfs,
-		if (rc) {
-			fprintf(stderr, "dfs mount on source failed: %d\n", rc);
-			D_GOTO(err_src_dfs, rc);
+		if (is_posix_copy) {
```
@dsikich Future improvement:
Similar to mpifileutils, it would be extensible to use something like `api in (DAOS, DFS)` instead of `is_posix_copy`.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2831,9 +2845,10 @@ fs_copy_hdlr(struct cmd_args_s *ap)
 	struct file_dfs		src_file_dfs = {0};
 	struct file_dfs		dst_file_dfs = {0};
-	struct fs_copy_args	fa = {0};
+	struct copy_args	fa = {0};
```
@dsikich Minor nitpick:
Since we changed `fs_copy_args` to `copy_args`, maybe we should also change `fa` to `ca` or something similar?

dsikich replied:
@daltonbohning I wasn't really looking at it that way. I kept the name since that instance of it is the "filesystem args," but I can go ahead and update it, doesn't really matter to me either way.

daltonbohning replied:
@dsikich That's okay, certainly not a big deal.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2948,6 +2963,467 @@ fs_copy_hdlr(struct cmd_args_s *ap)
 	return rc;
 }
 
+static int
+cont_copy_recx_single(daos_key_t *dkey,
```
@dsikich Future improvement:
Now that we have these functions here and in mpifileutils, we should consider centralizing them in some way.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2948,6 +2963,467 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+			d_iov_set(&diov, (void *)dkey, dkey_kds[j].kd_key_len);
+
+			/* enumerate and parse akeys */
+			rc = cont_copy_list_akeys(src_oh, dst_oh, diov);
```
@dsikich Question:
We keep looping here even if `cont_copy_list_akeys` failed. Should we continue?
If we do continue, then a future loop iteration might overwrite a non-zero `rc` with a `0`, and we would ultimately return 0, even though one of the `cont_copy_list_akeys` calls failed.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2948,6 +2963,467 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+	char	*cont_str = NULL;
+
+	/* check for DAOS pool/cont or UNS path */
+	rc = duns_resolve_path(arg, dattr);
+	if (rc != 0) {
+		pool_str = strtok_r(arg, "/", &saveptr);
+		rc = uuid_parse(pool_str, *pool);
+		if (rc == 0) {
+			cont_str = strtok_r(NULL, "/", &saveptr);
+			rc = uuid_parse(cont_str, *cont);
+			if (rc != 0) {
+				fprintf(stderr, "failed to parse cont "
+					"uuid: %d\n", rc);
+				D_GOTO(out, rc);
+			}
+		} else {
+			fprintf(stderr, "failed to parse uuid or "
+				"path: %d\n", rc);
+			D_GOTO(out, rc = EINVAL);
+		}
+	} else {
```
@dsikich Future improvement:
`duns_resolve_path` can also parse the pool/cont uuids as well. And you can set
`attr->da_no_prefix` to tell it whether or not the path has `daos:` prefix

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2948,6 +2963,467 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+	/* close src and dst pools, conts */
+	rc = copy_disconnect(is_posix_copy, &cp_args, &src_cp_type,
+			     &dst_cp_type);
+	if (rc != 0) {
+	}
```
@dsikich Should there be a message here?

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning reviewed (approved):
Thanks!

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2948,6 +2963,479 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		D_GOTO(out, rc);
+	}
+	/* only destroy snapshot if we created it, and it wasn't
+	 * passed in by the user */
```
(style)  Block comments use a trailing */ on a separate line

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4281/6/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/6/execution/node/66/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

dsikich commented:
@daltonbohning @mchaarawi @johannlombardi There has been some discussion between Dalton and I about when/if we should always delete the snapshot if an epoch was provided by the user? What we think is that we should pass DAOS_SNAP_OPT_CR | DAOS_SNAP_OPT_OIT when a snapshot is not provided by the user, and still call daos_cont_create_snap_opt with only the DAOS_SNAP_OPT_OIT when one is provided by the user? What is still not clear to me is if there is any way to tell if a snapshot already existed or that it was created. I just don't want to destroy a user's snapshot if that was not the intention. I will also hold off updating the code for this since I've updated it already, and want to hold off running the CI again until I'm sure my changes are correct.

daltonbohning commented:
> @daltonbohning @mchaarawi @johannlombardi There has been some discussion between Dalton and I about when/if we should always delete the snapshot if an epoch was provided by the user? What we think is that we should pass DAOS_SNAP_OPT_CR | DAOS_SNAP_OPT_OIT when a snapshot is not provided by the user, and still call daos_cont_create_snap_opt with only the DAOS_SNAP_OPT_OIT when one is provided by the user? What is still not clear to me is if there is any way to tell if a snapshot already existed or that it was created. I just don't want to destroy a user's snapshot if that was not the intention. I will also hold off updating the code for this since I've updated it already, and want to hold off running the CI again until I'm sure my changes are correct.

Something that isn't clear to me is that the documentation seems to indicate that taking a snapshot ALWAYS uses the latest epoch. This seems to also be in line with the way "daos cont create-snap" works. But if this is the case, then I'm confused about why the function `daos_cont_create_snap_opt` has the option `DAOS_SNAP_OPT_CR` to create at the current snapshot? Furthermore, the `epoch` parameter is just an "out" parameter, so passing it "in" doesn't seem to do anything:
https://github.com/daos-stack/daos/blob/943036c9216844176e343f2b0f7a85d85dab6416/src/include/daos_cont.h#L573-L589

dsikich commented:
@johannlombardi @mchaarawi @daltonbohning After doing some testing of these cases it seems there is no way currently to pass in an existing snapshot and create an OIT from it. The epoch as @daltonbohning pointed out is just an out param, so it creates a new snapshot/epoch each time you want to create an OIT. I tried just skipping creating a snapshot and going straight to daos_oit_open with an existing epoch, and that did not work. It seems the ability to use an already created snapshot to create an OIT is not supported yet, so I am going to remove that logic for now, unless I hear different.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning reviewed (dismissed):
👍  I think this is a good initial version of the tool.

daltonbohning commented:
@dsikich There are some merge conflicts; looks like from the `fs copy` changes that were just merged.

dsikich commented:
> @dsikich There are some merge conflicts; looks like from the `fs copy` changes that were just merged.

@daltonbohning thanks, working on getting this updated now.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

johannlombardi commented:
Could you please rebase?

mchaarawi commented on doc/man/man8/daos.8:
```diff
@@ -124,6 +126,11 @@ The \fBRESOURCE\fRs, respective \fBCOMMAND\fRs and \fBOPTION\fRs supported by \f
+.I container \fBOPTION\fRs (copy a container):
+	  \fB--src=\fRSTR      <pool uuid>/<cont uuid> | <UNS path>
+.br
+	  \fB--dst=\fRSTR      <pool uuid>/<cont uuid> | <UNS path>
```
is this always required?
let's say one wants to copy a container but wants us to generate the uuid for it.. same like when we call daos cont create. the uuid for the container is optional.

mchaarawi reviewed (changes requested):
other than the bugs to be fixed, there are some limitations currently imposed by this, and  we should make sure on the data mover WG that those are OK with the first landing. basically imposing the dkey/akey size limitations.
the other thing is that we don't fetch / set the max oid on the new container.
for that we need:
https://github.com/daos-stack/daos/pull/4421

also i don't understand why we need the destination container to exist. we should discuss this more.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2558,91 +2562,95 @@ fs_copy(struct file_dfs *src_file_dfs,
-		daos_cont_info_t *dst_cont_info,
-		struct duns_attr_t *src_dattr,
-		struct duns_attr_t *dst_dattr)
+copy_connect(bool is_posix_copy,
```
For the filesystem copy tool, it was mentioned that we should _**not**_ automatically create a destination container.
Unless I misunderstood?

mchaarawi replied:
but this is not the filesystem copy tool, right?
this is includes the generic tool.. let's discuss on skype.

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -2853,13 +2866,474 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+	}
+	memset(&anchor, 0, sizeof(anchor));
+	while (1) {
+		oids_nr = OID_ARR_SIZE;
```
@mchaarawi it is set to 8, the cont_list_objs_hdlr also uses it so I used the same one/size

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -2853,13 +2866,474 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+	/* check for DAOS pool/cont or UNS path */
+	rc = duns_resolve_path(arg, dattr);
+	if (rc != 0) {
+		pool_str = strtok_r(arg, "/", &saveptr);
```
@mchaarawi testing this... duns_resolve_path does not parse an input of the form /$pool/$cont it only returns the rel_path when the input has a prefix of daos:// from what I can tell. We could just use the daos:// prefix for them but in the case of the generic copy the type is always daos:// so I believe it is redundant?

mchaarawi replied:
you would need to set the da_no_prefix flag

dsikich replied:
@mchaarawi ok, got it working using the flag. I can remove this extra parse function I was using.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2853,13 +2866,474 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+				D_GOTO(out, rc);
+			}
+
+			/* if iod_size == 0 then this is a DAOS_IOD_ARRAY
```
you would need to set the da_no_prefix flag

mchaarawi replied:
@dsikich can you please verify with Jeff this. not sure he saw this.

daltonbohning replied:
I do see `iod_type` in the struct,  though I'm not sure if it's an "out" param:
https://github.com/daos-stack/daos/blob/caf75da13e02a30174882eeec16336967787831f/src/include/daos_obj.h#L192-L206

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -2853,13 +2866,474 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+			/* Print enumerated dkeys */
+			buf_len = snprintf(dkey, dkey_kds[j].kd_key_len + 1,
+					   "%s", ptr);
+			if (buf_len >= ENUM_KEY_BUF) {
```
@mchaarawi not really any special reason. When I was learning how to use the obj API I followed a lot of what the obj tests were doing, and from what I remember at the time they were using fixed size static buffers. Should I just create a buffer the size of the key length?

mchaarawi replied:
the test code works because we know the size of the keys there.
what you really need here is to start with a fixed size buffer (what you have is fine), but check that if you get the KEY2BIG error, to use a larger buffer instead of just failing.

dsikich replied:
@mchaarawi ok, I can update to doing it this way, thanks

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2528,92 +2532,202 @@ fs_copy(struct file_dfs *src_file_dfs,
-		struct duns_attr_t *src_dattr,
-		struct duns_attr_t *dst_dattr)
+dm_get_cont_type(struct dm_args *ca,
+	 	 char *sysname,
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2528,92 +2532,202 @@ fs_copy(struct file_dfs *src_file_dfs,
-		struct duns_attr_t *src_dattr,
-		struct duns_attr_t *dst_dattr)
+dm_get_cont_type(struct dm_args *ca,
+	 	 char *sysname,
+	   	 daos_cont_info_t *cont_info,
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
		 char *sysname,
		 daos_cont_info_t *cont_info,
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2528,92 +2532,202 @@ fs_copy(struct file_dfs *src_file_dfs,
-	/* check source pool/conts */
-	int rc = 0;
+	int                     rc = 0;
+	daos_prop_t*            prop = daos_prop_alloc(1);
```
(style)  "foo*            bar" should be "foo *bar"

daosbuild1 replied:
```suggestion
	daos_prop_t *prop = daos_prop_alloc(1);
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2528,92 +2532,202 @@ fs_copy(struct file_dfs *src_file_dfs,
+
+static int
+dm_connect(bool is_posix_copy,
+     	   struct file_dfs *src_file_dfs,
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
	   struct file_dfs *src_file_dfs,
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2528,92 +2532,202 @@ fs_copy(struct file_dfs *src_file_dfs,
+	/* check source pool/conts */
+	int			rc = 0;
+	struct duns_attr_t	dattr = {0};
+	
```
(style)  trailing whitespace

daosbuild1 replied:
```suggestion

```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2691,12 +2811,14 @@ fs_copy_disconnect(struct fs_copy_args *fa,
-fs_copy_parse_path(struct file_dfs *file, char *path,
-		   uuid_t *p_uuid, uuid_t *c_uuid)
+dm_parse_path(struct file_dfs *file, char *path,
+              uuid_t *p_uuid, uuid_t *c_uuid,
```
(style)  code indent should use tabs where possible

daosbuild1 replied:
```suggestion
	      uuid_t *p_uuid, uuid_t *c_uuid,
```

daosbuild1 replied:
(style)  code indent should use tabs where possible

daosbuild1 replied:
```suggestion
	      uuid_t *p_uuid, uuid_t *c_uuid,
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2824,13 +2952,588 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		(*iod).iod_type  = DAOS_IOD_ARRAY;
+		(*iod).iod_nr    = number;
+		(*iod).iod_recxs = recxs;
+		(*iod).iod_size  = size;  
```
(style)  trailing whitespace

daosbuild1 replied:
```suggestion
		(*iod).iod_size  = size;
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2824,13 +2952,588 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+	file_set_defaults_dfs(&dst_cp_type);
+
+	rc = dm_parse_path(&src_cp_type, ap->src, &ca.src_p_uuid,
+		           &ca.src_c_uuid, daos_no_prefix);
```
(style)  code indent should use tabs where possible

daosbuild1 replied:
```suggestion
			   &ca.src_c_uuid, daos_no_prefix);
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2824,13 +2952,588 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		ap->dst -= 1;
+		/* not considered an error yet since this could
+		 * be a UNS path, dm_connect will check this
+		 * when it attempts to do a uns_path_create */
```
(style)  Block comments use a trailing */ on a separate line

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2733,37 +2858,40 @@ fs_copy_hdlr(struct cmd_args_s *ap)
-	rc = fs_copy_parse_path(&src_file_dfs, src_str, &fa.src_p_uuid,
-				&fa.src_c_uuid);
+	rc = dm_parse_path(&src_file_dfs, src_str, &ca.src_p_uuid,
+			     &ca.src_c_uuid, daos_no_prefix);
```
```suggestion
			   &ca.src_c_uuid, daos_no_prefix);
```

daosbuild1 replied:
```suggestion
			   &ca.src_c_uuid, daos_no_prefix);
```

daosbuild1 replied:
```suggestion
			   &ca.src_c_uuid, daos_no_prefix);
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2733,37 +2858,40 @@ fs_copy_hdlr(struct cmd_args_s *ap)
-	rc = fs_copy_parse_path(&dst_file_dfs, dst_str, &fa.dst_p_uuid,
-				&fa.dst_c_uuid);
+	rc = dm_parse_path(&dst_file_dfs, dst_str, &ca.dst_p_uuid,
+			     &ca.dst_c_uuid, daos_no_prefix);
```
```suggestion
			   &ca.dst_c_uuid, daos_no_prefix);
```

daosbuild1 replied:
```suggestion
			   &ca.dst_c_uuid, daos_no_prefix);
```

daosbuild1 replied:
```suggestion
			   &ca.dst_c_uuid, daos_no_prefix);
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2824,13 +2952,588 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+
+static int
+cont_clone_recx_array(daos_key_t *dkey,
+		     daos_key_t *akey,
```
```suggestion
		      daos_key_t *akey,
```

daosbuild1 replied:
```suggestion
		      daos_key_t *akey,
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2824,13 +2952,588 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+
+static int
+cont_clone_list_akeys(daos_handle_t *src_oh,
+		     daos_handle_t *dst_oh,
```
```suggestion
		      daos_handle_t *dst_oh,
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2824,13 +2952,588 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+			if (akey == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+			len = snprintf(akey, akey_kds[j].kd_key_len + 1,
+					   "%s", ptr);
```
```suggestion
				       "%s", ptr);
```

daosbuild1 commented on src/utils/daos.c:
```diff
@@ -1413,6 +1430,11 @@ help_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 			ALL_BUT_CONT_CREATE_OPTS_HELP();
+		} else if (strcmp(argv[3], "clone") == 0) {
+			fprintf(stream,
+			"container options (clone):\n"
```
```suggestion
				"container options (clone):\n"
```

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4281/11/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/11/execution/node/68/log

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2528,92 +2532,202 @@ fs_copy(struct file_dfs *src_file_dfs,
-	/* check source pool/conts */
-	int rc = 0;
+	int                     rc = 0;
+	daos_prop_t*            prop = daos_prop_alloc(1);
```
(style)  "foo*            bar" should be "foo *bar"

daosbuild1 replied:
```suggestion
	daos_prop_t *prop = daos_prop_alloc(1);
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2528,92 +2532,202 @@ fs_copy(struct file_dfs *src_file_dfs,
+
+static int
+dm_connect(bool is_posix_copy,
+     	   struct file_dfs *src_file_dfs,
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
	   struct file_dfs *src_file_dfs,
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2528,92 +2532,202 @@ fs_copy(struct file_dfs *src_file_dfs,
+	/* check source pool/conts */
+	int			rc = 0;
+	struct duns_attr_t	dattr = {0};
+	
```
(style)  trailing whitespace

daosbuild1 replied:
```suggestion

```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2824,13 +2952,588 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		ap->dst -= 1;
+		/* not considered an error yet since this could
+		 * be a UNS path, dm_connect will check this
+		 * when it attempts to do a uns_path_create */
```
(style)  Block comments use a trailing */ on a separate line

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4281/17/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/17/execution/node/69/log

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/18/execution/node/69/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

dsikich commented:
@mchaarawi @daltonbohning I addressed the PR comments and updated how the destination container is handled. It only allows destination containers that do not exist already. Passing in the cuuid is optional now, but in the case of a UNS path, then the source container is used to create a new destination container. I also renamed the tool to "clone" vs "copy" as we discussed over Skype. I am going to wait for #4421 and #4403 to land so that I can pick up the patch @daltonbohning did to fix the coverity defects, and add the max oid allocation when the destination container is created with $4421.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2824,13 +2952,589 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		for (i = 0; i < number; i++) {
+			buf_len = recxs[i].rx_nr * size;
+			D_ALLOC(buf[i], buf_len);
+			if (buf[i] == NULL) {
+				D_FREE(buf);
+				D_GOTO(out, rc = -DER_NOMEM);
```
@dsikich I think if any of the `D_ALLOC(buf[i], buf_len)` calls fail, we also need to call `D_FREE` on each `buf[k]` for `k < i` (previous `i` values)

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,241 @@
-			D_GOTO(err_dst_root, rc);
+			if (ca->cont_layout == DAOS_PROP_CO_LAYOUT_POSIX) {
+				dfs_attr_t attr;
+				attr.da_props = props;
```
(style)  Missing a blank line after declarations

daosbuild1 replied:
```suggestion
				dfs_attr_t attr;

```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2770,19 +2936,20 @@
 	char			*dst_str = NULL;
 	size_t			src_str_len = 0;
-	size_t			dst_str_len = 0;
+	size_t		 	dst_str_len = 0;
```
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
	size_t			dst_str_len = 0;
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2770,19 +2936,20 @@
+	struct dm_args		ca = {0};
 	char			*name = NULL;
-	char			*dname = NULL;
+	char		    	*dname = NULL;
```
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
	char			*dname = NULL;
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2806,16 +2973,15 @@
-	rc = fs_copy_parse_path(&dst_file_dfs, dst_str, dst_str_len,
-				&fa.dst_p_uuid, &fa.dst_c_uuid);
+	rc = dm_parse_path(&dst_file_dfs, dst_str, dst_str_len,
+		      	   &ca.dst_p_uuid, &ca.dst_c_uuid, daos_no_prefix);
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
			   &ca.dst_p_uuid, &ca.dst_c_uuid, daos_no_prefix);
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3053,605 @@
+	char			*src_str = NULL;
+	char			*dst_str = NULL;
+	size_t			src_str_len = 0;
+	size_t		 	dst_str_len = 0;
```
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
	size_t			dst_str_len = 0;
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,241 @@
+			"container: %d\n", rc);
+	}
+
+	/* get alloced oid on source container to set on destination */
```
```suggestion
	/* get allocated oid on source container to set on destination */
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2729,12 +2891,13 @@
-fs_copy_parse_path(struct file_dfs *file, char *path, size_t path_len,
-		   uuid_t *p_uuid, uuid_t *c_uuid)
+dm_parse_path(struct file_dfs *file, char *path, size_t path_len,
+		      uuid_t *p_uuid, uuid_t *c_uuid, bool daos_no_prefix)
```
```suggestion
	      uuid_t *p_uuid, uuid_t *c_uuid, bool daos_no_prefix)
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2792,8 +2959,8 @@
-	rc = fs_copy_parse_path(&src_file_dfs, src_str, src_str_len,
-				&fa.src_p_uuid, &fa.src_c_uuid);
+	rc = dm_parse_path(&src_file_dfs, src_str, src_str_len,
+				&ca.src_p_uuid, &ca.src_c_uuid, daos_no_prefix);
```
```suggestion
			   &ca.src_p_uuid, &ca.src_c_uuid, daos_no_prefix);
```

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4281/21/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/21/execution/node/68/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional_Hardware_Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/22/execution/node/1219/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1710,11 +1719,12 @@ parse_filename_dfs(const char *path, char **_obj_name, char **_cont_name)
 	char	*fname = NULL;
 	char	*cont_name = NULL;
-	int	path_len = strlen(path) + 1;
+	int	path_len = 0;
```
no need to init to 0

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,241 @@ fs_copy(struct file_dfs *src_file_dfs,
+
+	if (prop == NULL) {
+		fprintf(stderr, "Failed to allocate prop (%d)", rc);
+		D_GOTO(out, rc);
```
rc here is 0.
should be D_GOTO(out, rc = -DER_NOMEM);

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,241 @@ fs_copy(struct file_dfs *src_file_dfs,
+		 uint64_t *dpe_val)
+{
+	int                     rc = 0;
+	daos_prop_t		*prop = daos_prop_alloc(1);
```
separate declaration and allocations please

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,241 @@ fs_copy(struct file_dfs *src_file_dfs,
+	}
+
+	/* get allocated oid on source container to set on destination */
+	rc = dm_get_cont_prop(ca->src_coh, sysname, src_cont_info,
```
you query the property twice now, once for the type and once for the oid.
you can do it in one get_prop call by allocating 2 properties. so you are adding an extra RPC vs you can just do it in one.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,241 @@ fs_copy(struct file_dfs *src_file_dfs,
+	rc = dm_get_cont_prop(ca->src_coh, sysname, src_cont_info,
+			      ca->cont_prop_layout, &ca->cont_layout);
+	if (rc != 0) {
+		fprintf(stderr, "failed to query source "
```
is there a reason you don't fail here and abort vs just print error and continue?

dsikich replied:
@mchaarawi well originally it was because the filesystem copy uses this connect call, and it would be ok for the query to fail if the source was not a DAOS container, but a POSIX file. I changed this to only query the source container if it is not of type POSIX.

mchaarawi replied:
if it is OK to fail then we should not print errors to stderr.
but i don't get why get prop would fail and why it would be ok to fail, whatever the container type is

dsikich replied:
@mchaarawi well I mean in this case if src_mfu_file->type == POSIX it is not a DAOS container, the source would be coming from the POSIX filesystem. I updated this to check if that was the source, so it skips the container query in that case.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,241 @@ fs_copy(struct file_dfs *src_file_dfs,
+	rc = dm_get_cont_prop(ca->src_coh, sysname, src_cont_info,
+			      ca->cont_prop_oid, &ca->cont_oid);
+	if (rc != 0) {
+		fprintf(stderr, "failed to query source "
```
is there a reason you don't fail here and abort vs just print error and continue?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,241 @@ fs_copy(struct file_dfs *src_file_dfs,
+			rc = duns_create_path(ca->dst_poh, path,
+					      &dattr);
+			if (rc != 0) {
+				fprintf(stderr, "please provide a destination "
```
style: im not sure we ever use "please" in our output.
maybe just print what the error is rather than a command to the user.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,241 @@ fs_copy(struct file_dfs *src_file_dfs,
+		rc = daos_cont_open(ca->dst_poh, ca->dst_c_uuid,
+				    DAOS_COO_RW, &ca->dst_coh,
+				    dst_cont_info, NULL);
 		if (rc != 0) {
```
TBH i'm not understanding what is going on here.
you need some comments in the code to explain the logic.
why do we attempt to create on every error?

dsikich replied:
@mchaarawi the first check is if the dst pool uuid is null, and if it is then this is a UNS destination path, and I have to copy the src pool uuid into the dst pool uuid, then create the UNS path on the destination. I first try a daos_cont_open because it could be a filesystem copy, and it is not required that the dest container doesn't already exist, and if it fails then I need to create the destination container (whether it is for the filesystem or clone copy). I then have to try to open the container again after it is created. I am adding some comments to hopefully make this more clear.

mchaarawi replied:
the issue is that you are issuing a create for **any** type of error, not just DER_NOEXIST errors!
and yes some comments are needed here to help understand what is going on.

dsikich replied:
@mchaarawi ah yeah I somehow missed that! I will update that, thanks.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,241 @@ fs_copy(struct file_dfs *src_file_dfs,
+					    dst_cont_info, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to open container: "
+					"%d\n", rc); D_GOTO(err_dst_root, rc);
```
move goto to a new line

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3053,605 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+	daos_epoch_t		epoch;
+	struct			dm_args ca = {0};
+	bool			is_posix_copy = false;
+	daos_handle_t		oh = DAOS_HDL_INVAL;
```
is there a reason to init this, and all other things like it?

dsikich replied:
@mchaarawi I think (but can't remember exactly) because at some point I got an error in one of the builds when submitting to Jenkins that the handles were not initialized in some cases, I can try and remove these and see if I still get it.

mchaarawi replied:
oh im not suggesting you remove everything here and retry.
but usually you never need to initialize handles unless you are misusing them.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3053,605 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+				       "%s", ptr);
+			if (len > key_buf_len) {
+				fprintf(stderr, "dkey buffer is too small\n");
+				D_GOTO(out, rc = 1);
```
leaked dkey

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,267 @@ fs_copy(struct file_dfs *src_file_dfs,
+	for (i = 0; i < size; i++) {
+		prop->dpp_entries[i].dpe_type = dpe_types[i];
+	}
+	//prop->dpp_entries[0].dpe_type = dpe_type;
```
(style)  do not use C99 // comments

daosbuild1 replied:
```suggestion
	/* prop->dpp_entries[0].dpe_type = dpe_type; */
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,267 @@ fs_copy(struct file_dfs *src_file_dfs,
+	for (i = 0; i < size; i++) {
+		dpe_vals[i] = prop->dpp_entries[i].dpe_val;
+	}
+	//*dpe_val = prop->dpp_entries[0].dpe_val;
```
(style)  do not use C99 // comments

daosbuild1 replied:
```suggestion
	/* *dpe_val = prop->dpp_entries[0].dpe_val; */
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,267 @@ fs_copy(struct file_dfs *src_file_dfs,
+		dpe_vals[i] = prop->dpp_entries[i].dpe_val;
+	}
+	//*dpe_val = prop->dpp_entries[0].dpe_val;
+	daos_prop_free(prop);
```
(style)  Block comments use * on subsequent lines

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4281/25/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/25/execution/node/68/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Scan CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/28/execution/node/385/log

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/28/execution/node/410/log

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3078,606 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		 */
+		rc = 0;
+	} else {
+		rc = daos_pool_connect(ca.dst_p_uuid, ap->sysname,
+				       DAOS_PC_RW, &ca.dst_poh, NULL, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to connect to destination "
+				"pool: %d\n", rc);
+			D_GOTO(out, rc);
+		}
+		/* make sure this destination container doesn't exist already,
+		 * if it does, exit
+		 */
+		rc = daos_cont_open(ca.dst_poh, ca.dst_c_uuid, DAOS_COO_RW,
```
@dsikich Doesn't `dm_connect` below connect to the pool and cont?

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daltonbohning commented:
@dsikich Running with no arguments
```
[dbohninx@boro-43 dbohninx]$ daos cont clone
...
Segmentation fault
```
Running with valgrind:
```
[dbohninx@boro-43 dbohninx]$ valgrind daos cont clone
...
==6523== Invalid read of size 1
==6523==    at 0x4C2D0F2: strlen (vg_replace_strmem.c:461)
==6523==    by 0x416DD5: cont_clone_hdlr (daos_hdlr.c:3497)
==6523==    by 0x40949C: cont_op_hdlr (daos.c:989)
==6523==    by 0x405378: main (daos.c:1593)

```

daltonbohning commented:
@dsikich Running with a src but no dst
```
[dbohninx@boro-43 dbohninx]$ daos cont clone --src /$pool1/$cont1
...
Segmentation fault
```
```
[dbohninx@boro-43 dbohninx]$ valgrind daos cont clone --src /$pool1/$cont1
...
==6764== Invalid read of size 1
==6764==    at 0x4C2D0F2: strlen (vg_replace_strmem.c:461)
==6764==    by 0x416E64: cont_clone_hdlr (daos_hdlr.c:3510)
==6764==    by 0x40949C: cont_op_hdlr (daos.c:989)
==6764==    by 0x405378: main (daos.c:1593)
```

daltonbohning commented on src/utils/daos.c:
```diff
@@ -971,38 +971,50 @@ static int
-		 *
-		 * Firtly check for a unified namespace entry point, then if
-		 * that isn't detected then check for dfuse backing the
-		 * path, and print pool/container/oid for the path.
+		 * Firtly check for a unified namespace entry point,
+		 * then if that isn't detected then check for dfuse
+		 * backing the path, and print pool/container/oid
+		 * for the path.
 		 */
 		rc = duns_resolve_path(ap->path, &dattr);
 		if (rc) {
-
 			rc = call_dfuse_ioctl(ap->path, &il_reply);
 			if (rc != 0) {
-				fprintf(stderr, "could not resolve pool, "
-					"container by path: %d %s %s\n",
+				fprintf(stderr, "could not resolve "
+					"pool, container by "
+					"path: %d %s %s\n",
```
@dsikich Minor format:
(Only if pushing again)
Since this was shifted and then unshifted, we could just revert these lines to exactly what they were originally to simplify the diffs, and remove the need for extra lines.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,266 @@ fs_copy(struct file_dfs *src_file_dfs,
+	int                     i = 0;
+	daos_prop_t		*prop = NULL;
+
+	prop = daos_prop_alloc(2);
```
@dsikich Should this instead be
```suggestion
	prop = daos_prop_alloc(size);
```

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,266 @@ fs_copy(struct file_dfs *src_file_dfs,
+					"destination pool: %d\n", rc);
+				D_GOTO(out, rc);
+			}
+		/* if the dst pool uuid is null that means that this
+		 * is a UNS destination path, so we copy the source
+		 * pool uuid into the destination and try to connect
+		 * again
+		 */
```
@dsikich Are we connecting to the same pool twice? Maybe that's okay, but just checking.

dsikich replied:
@daltonbohning possibly, and I believe that is ok. It simplifies things quite a bit to just get our own hdlr for the src and destination each time.

daltonbohning replied:
@dsikich Makes sense!

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,266 @@ fs_copy(struct file_dfs *src_file_dfs,
+		rc = daos_cont_open(ca->dst_poh, ca->dst_c_uuid,
+				    DAOS_COO_RW, &ca->dst_coh,
+				    dst_cont_info, NULL);
+		if (rc == -DER_NONEXIST) {
```
@dsikich I think we need an else for this, right?
Something like:
```
if (rc == -DER_NONEXIST) {
    // create it and try to open again
} else {
    // error
}
```

dsikich replied:
@daltonbohning well that isn't really an error, since if it successfully opens then it isn't an error, but I suppose if rc != 0 at that point it is so I can check

daltonbohning replied:
@dsikich Yes, sorry, that's what I meant to say. If rc is 0 and it's also not -DER_NONEXIST, then that is a real error.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3078,606 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+	 */
+	if (rc != 0) {
+		uuid_generate(ca.dst_c_uuid);
+		ap->dst += 1;
```
@dsikich This += assumes the path starts with '/'.
But what if it doesn't?

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3078,606 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		/* make sure this destination container doesn't exist already,
+		 * if it does, exit
+		 */
+		rc = daos_cont_open(ca.dst_poh, ca.dst_c_uuid, DAOS_COO_RW,
```
@dsikich Doesn't dm_connect below connect to the pool and cont?

dsikich replied:
@daltonbohning yes, but I figured there is no reason to enter that function and connect to the src if the destination already exists, since that is an error. Plus, fs copy uses dm_connect, and this is not an error for fs copy.

daltonbohning replied:
@dsikich What I mean is that if we connect to the dst pool here, aren't we connecting to the dst pool again in dm_connect? When we connect the second time, the first handle is still open, right?

dsikich replied:
@daltonbohning hmm well I suppose in dm_connect I only need to try and connect to the pool again if the dst uuid is null, but I think I do this because if this is a filesystem copy then the pool would not already be open. I can close the pool before entering dm_connect though, since doing it the other way will not work for fs copy.

daltonbohning replied:
@dsikich Maybe `daos_handle_is_valid` could be useful in dm_connect to only connect if the handle is NOT valid. I.e. we haven't yet connected

dsikich replied:
@daltonbohning it looks like that should work, I can add a check for a valid handle before calling connect on the destination

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3078,606 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		}
+	}
+	D_GOTO(out_oit, rc);
+err_obj:
+	rc = daos_obj_close(oh, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to close source object: %d\n", rc);
+	}
+	rc = daos_obj_close(dst_oh, NULL);
+	if (rc != 0) {
+		fprintf(stderr, "failed to close destination object: %d\n", rc);
+	}
+err_dst:
+	rc = daos_obj_close(oh, NULL);
```
@dsikich Maybe it's okay, but if we goto err_obj, then both err-obj and err_dst try to close `oh`

dsikich replied:
@daltonbohning I don't know if it is not ok, but not really sure a way around that for that case.

daltonbohning replied:
@dsikich Could we just remove it from err_obj and let it fall through to err_dst?

dsikich replied:
@daltonbohning yeah I guess that would close them both in if both objects are already opened

daltonbohning reviewed (commented):
@dsikich There's still an issue where large files are not being copied entirely.
```
[dbohninx@boro-43 dbohninx]$ ls -l dfuse/$pool1/$cont1
total 1048576
-rw-rw-r-- 1 dbohninx dbohninx 1073741824 Feb 18 21:54 large_file

[dbohninx@boro-43 dbohninx]$ daos cont clone --src /$pool1/$cont1 --dst /$pool1
Successfully copied to destination container 10d13ec1-2afc-4648-91bc-f22d369f6c6a
[dbohninx@boro-43 dbohninx]$ cont2=10d13ec1-2afc-4648-91bc-f22d369f6c6a

[dbohninx@boro-43 dbohninx]$ ls -l dfuse/$pool1/$cont2
total 1047552
-rw-rw-r-- 1 dbohninx dbohninx 1072693248 Feb 18 21:54 large_file
```

Notice the different in file size. This is for a POSIX container, which uses the daos_array API for files, which I believe creates many dkeys/akeys automatically.

daltonbohning commented:
> @dsikich There's still an issue where large files are not being copied entirely.
> ...
>

After some debugging, the problem is that we're using `snprintf` to copy each dkey to the dkey buffer. Replacing the `snprintf` with an equivalent `memcpy` seems to work.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3078,606 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+			D_ALLOC(akey, key_buf_len);
+			if (akey == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+			len = snprintf(akey, akey_kds[j].kd_key_len + 1,
```
@dsikich We need to use `memcpy(akey, ptr, akey_kds[j].kd_key_len)`

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3078,606 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+			D_ALLOC(dkey, key_buf_len);
+			if (dkey == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+			len = snprintf(dkey, dkey_kds[j].kd_key_len + 1,
```
@dsikich We need to use `memcpy(dkey, ptr, dkey_kds[j].kd_key_len)`

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,274 @@ fs_copy(struct file_dfs *src_file_dfs,
+		if (!uuid_is_null(ca->dst_p_uuid)) {
+			if (!daos_handle_is_valid(ca->dst_poh)) {
+				rc = daos_pool_connect(ca->dst_p_uuid, sysname,
+		       	       			       DAOS_PC_RW, &ca->dst_poh,
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,274 @@ fs_copy(struct file_dfs *src_file_dfs,
+		if (!uuid_is_null(ca->dst_p_uuid)) {
+			if (!daos_handle_is_valid(ca->dst_poh)) {
+				rc = daos_pool_connect(ca->dst_p_uuid, sysname,
+		       	       			       DAOS_PC_RW, &ca->dst_poh,
+				       		       NULL, NULL);
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
						       DAOS_PC_RW, &ca->dst_poh,
						       NULL, NULL);
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,274 @@ fs_copy(struct file_dfs *src_file_dfs,
+				"%d\n", rc);
+			D_GOTO(err_dst_root, rc);
+
+		}	      	
```
(style)  trailing whitespace
(style)  please, no space before tabs

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3086,595 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+			uuid_parse(ca.dst, ca.dst_p_uuid);
+			ap->dst -= 1;
+			/* not considered an error yet since this could
+		 	* be a UNS path, dm_connect will check this
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3086,595 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+			ap->dst -= 1;
+			/* not considered an error yet since this could
+		 	* be a UNS path, dm_connect will check this
+		 	* when it attempts to do a uns_path_create
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3086,595 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+			uuid_parse(ca.dst, ca.dst_p_uuid);
+			ap->dst -= 1;
+			/* not considered an error yet since this could
+		 	* be a UNS path, dm_connect will check this
+		 	* when it attempts to do a uns_path_create
+		 	*/
```
(style)  code indent should use tabs where possible
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
			* be a UNS path, dm_connect will check this
			* when it attempts to do a uns_path_create
			*/
```

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,274 @@ fs_copy(struct file_dfs *src_file_dfs,
+			fprintf(stderr, "failed to open container: "
+				"%d\n", rc);
+			D_GOTO(err_dst_root, rc);
+
```
```suggestion
		}
```

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4281/30/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/30/execution/node/72/log

daosbuild1 commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,274 @@ fs_copy(struct file_dfs *src_file_dfs,
-			fprintf(stderr, "dfs mount on destination "
-			"failed: %d\n", rc);
-			D_GOTO(err_dst_dfs, rc);
+		}	      	
```
(style)  trailing whitespace
(style)  please, no space before tabs

daosbuild1 replied:
```suggestion
		}
```

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-4281/33/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/33/execution/node/75/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/33/execution/node/376/log

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-4281/33/execution/node/348/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

dsikich commented:
> > @dsikich There's still an issue where large files are not being copied entirely.
> > ...
> 
> After some debugging, the problem is that we're using `snprintf` to copy each dkey to the dkey buffer. Replacing the `snprintf` with an equivalent `memcpy` seems to work.

@daltonbohning this has been updated to use memcpy

daltonbohning commented on src/utils/daos.c:
```diff
@@ -986,7 +986,11 @@ cont_op_hdlr(struct cmd_args_s *ap)
 	 */
 	if (op == CONT_CLONE) {
-		rc = cont_clone_hdlr(ap);
+		if (ap->src != NULL && ap->dst != NULL) {
+			rc = cont_clone_hdlr(ap);
+		} else {
+			rc = RC_PRINT_HELP;
+		}
```
@dsikich This doesn't give the user an indication of what went wrong. To be similar to other tools, maybe:
```suggestion
		if (ap->src != NULL && ap->dst != NULL) {
			rc = cont_clone_hdlr(ap);
		} else {
			fprintf(stderr, "Source and destination required\n");
			rc = RC_PRINT_HELP;
		}
```

daltonbohning reviewed (approved):
@dsikich Just a minor comment. Otherwise, the "happy path" seems to be in order.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2566,95 +2576,274 @@ fs_copy(struct file_dfs *src_file_dfs,
+		/* only set max oid on created dest containers
+		 * if this is a non-posix copy
+		 */
+		if (!is_posix_copy) {
```
it's a little confusing to readers the difference between:
is_posix_copy vs src_file_dfs->type != POSIX
I would go the Dalton's suggestion with API name.
please do this in a follow on patch. doesn't have to be this one.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3086,595 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+	char		*key_buf = NULL;
+	daos_size_t	key_buf_len = 0;
+
+	D_ALLOC(small_key, ENUM_DESC_BUF);
```
future improvement:
for small keys buffer, it's probably better to just have that on the stack to avoid too many allocations.

daltonbohning replied:
@mchaarawi Something I've thought about is the possibility of passing buffers _**to**_ functions like this - maybe in a struct. Then, we could start with a small alloc that is done a single time and passed with each call. In this function, we could realloc as needed for larger buffers.
Of course, there are some caveats, but what do you think about this general idea?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -2887,6 +3086,595 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		rc = daos_obj_list_akey(*src_oh, DAOS_TX_NONE, &diov,
+					&akey_number, akey_kds,
+					&sgl, &akey_anchor, NULL);
+		if (rc == -DER_KEY2BIG) {
```
future improvement:
this should be a continued try with larger buffer if KEY2BIG scenario, because we don't impose a limitation on the key buffer size (for akeys and dkeys).
but this is good enough for a first cut.
